src/core/server/http/router/validator/validator.ts

```diff
@@ -0,0 +1,244 @@
+  /**
+   * Set the `unsafe` config to avoid running some additional internal *safe* validations on top of your custom validation
+   * @public
+   */
+  unsafe?: {
+    params?: boolean;
+    query?: boolean;
+    body?: boolean;
+  };
+}
+
+/**
+ * Route validations config and options merged into one object
+ * @public
+ */
+export type RouteValidatorFullConfig<P, Q, B> = RouteValidatorConfig<P, Q, B> &
+  RouteValidatorOptions;
+
+/**
+ * Route validator class to define the validation logic for each new route.
+ *
+ * @private
+ */
+export class RouteValidator<P = {}, Q = {}, B = {}> {
+  public static from<P = {}, Q = {}, B = {}>(
+    opts: RouteValidator<P, Q, B> | RouteValidatorFullConfig<P, Q, B>
+  ) {
+    if (opts instanceof RouteValidator) {
+      return opts;
+    }
+    const { params, query, body, ...options } = opts;
+    return new RouteValidator({ params, query, body }, options);
+  }
+
+  constructor(
+    private readonly config: RouteValidatorConfig<P, Q, B>,
+    private readonly options: RouteValidatorOptions = {}
+  ) {}
+
+  /**
+   * Get validated URL params
+   * @private
+   */
+  public getParams(data: unknown, namespace?: string): Readonly<P> {
+    return this.validate(this.config.params, this.options.unsafe?.params, data, namespace);
+  }
+
+  /**
+   * Get validated query params
+   * @private
+   */
+  public getQuery(data: unknown, namespace?: string): Readonly<Q> {
+    return this.validate(this.config.query, this.options.unsafe?.query, data, namespace);
+  }
+
+  /**
+   * Get validated body
+   * @private
+   */
+  public getBody(data: unknown, namespace?: string): Readonly<B> {
+    return this.validate(this.config.body, this.options.unsafe?.body, data, namespace);
+  }
+
+  /**
+   * Has body validation
+   * @private
+   */
+  public hasBody(): boolean {
+    return typeof this.config.body !== 'undefined';
+  }
+
+  private validate<T>(
+    validationRule?: RouteValidationSpec<T>,
+    unsafe?: boolean,
+    data?: unknown,
+    namespace?: string
+  ): RouteValidatedValue<typeof validationRule> {
+    if (typeof validationRule === 'undefined') {
+      return {};
+    }
+    let precheckedData = this.preValidateSchema(data).validate(data, {}, namespace);
+
+    if (unsafe !== true) {
+      precheckedData = this.safetyPrechecks(precheckedData, namespace);
+    }
+
+    const customCheckedData = this.customValidation(validationRule, precheckedData, namespace);
+
+    if (unsafe === true) {
+      return customCheckedData;
+    }
+
+    return this.safetyPostchecks(customCheckedData, namespace);
+  }
+
+  private safetyPrechecks<T>(data: T, namespace?: string): T {
+    // We can add any pre-validation safety logic in here
+    return data;
+  }
+
+  private safetyPostchecks<T>(data: T, namespace?: string): T {
+    // We can add any post-validation safety logic in here
+    return data;
+  }
+
+  private customValidation<T>(
+    validationRule?: RouteValidationSpec<T>,
```

Comment:
I think you can make this non-optional and remove the `if undefined` below

Reply:
The thing is the `validationRule` might be not set. Aka: some routes check `body` but not `params` (and the other way around), so you get `query` and `params` as `undefined` and the current logic simply returns those 2 properties as `{}`. 

I might prefer not to return anything at all. But our contract is to always return those params, either the validated value or `{}` if no validation is provided. The current `master` implementation does it in a very repetitive way: https://github.com/elastic/kibana/blob/2b6ef5c2bb1c14b3fdb1b2bc1b2cad06221d561b/src/core/server/http/router/request.ts#L106-L129

Reply:
That makes sense, but I think the defined-ness is already checked in the function that calls this one. It doesn't appear that this method is ever called with an undefined validationRule.

Reply:
Oh! Eagle eye!